opal/corelib/range.rb

```diff
@@ -38,6 +38,36 @@ def cover?(value)
   def each(&block)
     return enum_for :each unless block_given?
 
+    %x{
+      var i, limit, value;
+
+      if (#@begin.$$is_number && #@end.$$is_number) {
+        if (#@begin % 1 !== 0 || #@end % 1 !== 0) {
+          #{raise TypeError, "can't iterate from Float"}
+        }
+
+        for (i = #@begin, limit = #@end + #{@exclude ? 0 : 1}; i < limit; i++) {
+          value = block(i);
+          if (value === $breaker) { return $breaker.$v; }
+        }
+
+        return self;
+      }
+
+      if (#@begin.$$is_string && #@end.$$is_string) {
+        value = #{@begin.upto(@end, @exclude, &block)};
+
+        // The following is a bit hackish: we know that
+        // String#upto normally returns self, but may
+        // return a different value if there's a `break`
+        // statement in the supplied block. We need to
+        // propagate this `break` value here, so we
+        // test for equality with `@begin` string to
+        // determine the return value:
```

Comment:
I would love to hear your thoughts on this. What's the current "best practice" for propagating the `break` value, e.g. in this case where we delegate to another method which takes a block, so we are passing the block down the line, but we still need to know if `break` was used somewhere (anywhere, arbitrarily deeply) down the line... :confused: